pkg/controller/servicemesh/controlplane/reconciler.go

```diff
@@ -208,34 +216,150 @@ func (r *ControlPlaneReconciler) Reconcile() (reconcile.Result, error) {
+		}
+		templateContent = defaultTemplateContent
+	}
+
+	var template v1.ServiceMeshControlPlane
+	if err = yaml.Unmarshal(templateContent, &template); err != nil {
+		return v1.ControlPlaneSpec{}, fmt.Errorf("failed to parse template %s contents: %s", name, err)
+	}
+	return template.Spec, nil
+}
+
+//renderSMCPTemplates traverses and processes all of the references templates
+func (r *ControlPlaneReconciler) recursivelyApplyTemplates(smcp v1.ControlPlaneSpec, visited map[string]struct{}) (v1.ControlPlaneSpec, error) {
+	if smcp.Template == "" {
+		return smcp, nil
+	}
+	r.Log.Info(fmt.Sprintf("processing smcp template %s", smcp.Template))
+
+	if _, ok := visited[smcp.Template]; ok {
+		return smcp, fmt.Errorf("SMCP templates form cyclic dependency. Cannot proceed")
+	}
+
+	template, err := r.getSMCPTemplate(smcp.Template)
+	if err != nil {
+		return smcp, err
+	}
+
+	template, err = r.recursivelyApplyTemplates(template, visited)
+	if err != nil {
+		r.Log.Info(fmt.Sprintf("error rendering SMCP templates: %s\n", err))
+		return smcp, err
+	}
+
+	visited[smcp.Template] = struct{}{}
+
+	smcp.Istio = mergeValues(smcp.Istio, template.Istio)
+	smcp.ThreeScale = mergeValues(smcp.ThreeScale, template.ThreeScale)
+	return smcp, nil
+}
+
+func (r *ControlPlaneReconciler) applyTemplates(smcpSpec v1.ControlPlaneSpec) (v1.ControlPlaneSpec, error) {
+	r.Log.Info("updating servicemeshcontrolplane with templates")
+	if smcpSpec.Template == "" {
+		smcpSpec.Template = smcpDefaultTemplate
+		r.Log.Info("No template provided. Using default")
+	}
+
+	spec, err := r.recursivelyApplyTemplates(smcpSpec, make(map[string]struct{}, 0))
+	r.Log.Info(fmt.Sprintf("finished updating ServiceMeshControlPlane: %+v", spec))
+
+	return spec, err
+}
+
+func (r *ControlPlaneReconciler) validateSMCPSpec(spec v1.ControlPlaneSpec) error {
+	if spec.Istio == nil {
+		return fmt.Errorf("ServiceMeshControlPlane missing Istio section")
+	}
+
+	if _, ok := spec.Istio["global"].(map[string]interface{}); !ok {
+		return fmt.Errorf("ServiceMeshControlPlane missing global section")
+	}
+	return nil
+}
+
 func (r *ControlPlaneReconciler) renderCharts() error {
-	r.Log.Info("rendering helm charts")
-	allErrors := []error{}
-	var err error
-	var threeScaleRenderings map[string][]manifest.Manifest
+	//Generate the spec
+	r.Status.LastAppliedConfiguration = r.Instance.Spec
+
+	spec, err := r.applyTemplates(r.Status.LastAppliedConfiguration)
+	if err != nil {
+		r.Log.Error(err, "warning: failed to apply ServiceMeshControlPlane templates: %s", err)
+
+		return err
+	}
+	r.Status.LastAppliedConfiguration = spec
+
+	if err := r.validateSMCPSpec(r.Status.LastAppliedConfiguration); err != nil {
+		return err
+	}
 
-	if globalValues, ok := r.Instance.Spec.Istio["global"].(map[string]interface{}); ok {
+	if globalValues, ok := r.Status.LastAppliedConfiguration.Istio["global"].(map[string]interface{}); ok {
 		globalValues["operatorNamespace"] = r.OperatorNamespace
-	} else {
```

Comment:
Why did you remove this?  This is required for the cni charts.  If we can't set this, then we can't proceed with the rendering/install.

Reply:
..,or we should proceed as below and create a map for the `global` field.

Reply:
@rcernich I moved validation of the global section being present into the SMCP validation (312).

Reply:
r.Status.LastAppliedConfiguration.Istio["global"]

still has the same problem as the previous code if the above layers aren't checked. If .Istio is nil, then attempting to access the global field will crash.

And the global section not being present at all is likely an issue even outside of setting the operatornamespace. I.e. what tag do you use?